Modularize Reader state

sgomes opened the pull request:
## 📢Let the grand modularization of state of 2020... commence! 📢

Ahem.

This is the first of what are likely to be many PRs attempting to introduce some modularization of state to Calypso. This one deals with Reader, specifically. As the first of the PRs, it also introduces the mechanism through which modularization will happen.

 See p4TIVU-9lM-p2

### How state modularization works

We don't make any changes to reducers, selectors, or actions themselves for this to work, nor do we change the Redux architecture in any way. The approach we take builds on top of Redux by adding a new, developer-enforced requirement: creating static dependencies between actions/selectors and the reducers that control the parts of state they're interested in.

For this to work, reducers first need to be able to register themselves. That's done with the addition of a global Redux store handler, which exposes a registration method for reducers. It coordinates registration with store creation through the use of a synchronous queue.

Once a reducer is able to register itself, we can add the registration as a side effect to its main module. Other modules that make use of that reducer's corresponding state, such as action and selector modules, can then simply perform a naked import of the reducer module, to ensure that it's loaded (and thus registered) in time for their methods to be used.

This way, we achieve modularization through the dependency graph, and get code splitting for free through the regular build process. The code moves from upfront chunks that get loaded on every entry point to only the relevant chunks, removing the need for every entry point to load all of state, and thus reducing the amount of JS the end user needs to load, parse, and execute on any given route.

### Tradeoff for consideration

Redux state is a tree, and we could arguably set modularization at any level in that tree. Finer granularity gets us more potential savings, at the cost of more development overhead: it's easier to add a single import for the reader reducer than it is to add one for reader conversations, another for reader feeds, etc.

In this PR, I opted for a very coarse level of granularity, by splitting at the top-level key. I expect this to offer a good tradeoff in terms of getting some significant gains while keeping things simple, but I'd like to hear your thoughts.

### Changes proposed in this Pull Request

* Add global Redux store handler
* Set global Redux store on boot (for the `entry-main` and `entry-login` entry points)
* Remove Reader reducer instantiation from `entry-main` boot
* Add naked imports of Reader reducer to all Reader-related actions and selectors
* Minor lint fixes

I split this PR in three separate commits to facilitate reviewing the different aspects separately.

### Testing instructions

* Install [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).
* Open DevTools and switch to the Redux tab.
* Open `/home`.
* Click `State` on the top right to ensure you see all state, and not just the diff for the last action: 
![image](https://user-images.githubusercontent.com/409615/73774406-d3e87d80-477b-11ea-9f59-3e42cc00101f.png)
* Analyze the tree and note that there's no `reader` key, even if you expand the list of keys. This indicates that the Reader state hasn't been loaded yet.
* Click `Reader` on the masterbar.
* Verify that a `reader` key is added with the Reader state. This indicates that the Reader state has now been loaded.
* Verify that Reader works normally. This indicates that I didn't mess up.

sgomes committed:
Add global redux store handler.

sgomes committed:
Set global redux store on boot.

sgomes committed:
Modularize Reader reducer.

This consists of:
- Removing the Reader reducer from store creation on boot
- Adding a side-effect to the reducer module so that it registers itself
- Adding a naked import of the reducer module on all selector and
   action modules that read from or write to Reader state.

sgomes commented on client/state/redux-store.ts:
```diff
@@ -0,0 +1,43 @@
+	reducerRegistrationQueue = [];
+}
+
+export function getStore() {
```
I added this method because I tend to think symmetrically, but I'm not seeing a good use for it anymore. Should I remove it?

dmsnell replied:
did you answer your own question?

sgomes replied:
I suppose I was looking for validation more than an answer 🙂 I'll take your comment as one of the former 😄

blowery commented:
Is this introducing a number of new cycles into the module graph? If so, do they matter?

sgomes commented:
> Is this introducing a number of new cycles into the module graph? If so, do they matter?

I don't think that there are new cycles being introduced, since we're just creating a dependency going from actions/selectors to reducers. I don't think that reducers tend to depend on actions (only action name constants) nor selectors, so there should be no cycles being introduced.

Furthermore, even if cycles were being introduced, that's not necessarily an issue, particularly when we're just relying on a side-effect like we are here.

sgomes commented:
Rebased on master due to mass lint fixes in #39260.

sgomes committed:
Remove `getStore` method.

tyxla reviewed (commented):
Really nice work, this is so exciting to see 👏 

I'm getting this warning in the console:

![](https://cldup.com/8LxPhxBSv1.png)

which makes sense - at the time of the initial state hydration on the server-side, there the reader isn't a valid reducer subtree. 

Have we thought about an elegant way to go around this?

sgomes commented:
> I'm getting this warning in the console:

That's interesting, I can't seem to reproduce that error. Is there a specific set of steps you followed to get it?

tyxla commented:
@sgomes I assume you're unable to reproduce because you deleted your local store at least once while working on this 😉 

To repro, you could checkout master and build it, then run the app and go through the reader so it could store that tree for you. Then, going back to this branch and trying again should bring that error for you, as the `reader` key will be there in the local store, but the reducer won't exist as a valid reducer subtree in the initial state at boot time. Let me know if that works.

sgomes commented:
@tyxla Thank you! I'll see what I can do 👍

sgomes committed:
Avoid console errors when deserializing.

Redux complains when the state being deserialized contains more
properties than the reducers know about. In order to avoid these errors,
we store the list of keys for a reducer as a property of the reducer,
and we filter out the stored state for the expected properties when
deserializing.

sgomes commented:
@tyxla: How does the solution in eb9283459fcb0b0de947b79672d5b0f6055f1132 look?

sgomes commented:
Looks like matticbot is producing incorrect numbers; these look like the total sizes, not the delta. Shame, because I was looking forward to seeing what the delta looked like 😕

tyxla commented:
> @tyxla: How does the solution in [eb92834](https://github.com/Automattic/wp-calypso/commit/eb9283459fcb0b0de947b79672d5b0f6055f1132) look?

This solves the problem in question 👍 

However, I think it conceals a more concerning problem: since we're picking only the monolithic old-school reducers here, won't this obliterate any persistence for the modularized reducers?

tyxla commented:
> Looks like matticbot is producing incorrect numbers; these look like the total sizes, not the delta. Shame, because I was looking forward to seeing what the delta looked like 😕

Yeah 😞 I'm assuming it's because it can't calculate the total size of `master` for some reason, but I'm not familiar with ICFY at all - @jsnajdr any idea what might be going on?

sgomes commented:
> However, I think it conceals a more concerning problem: since we're picking only the monolithic old-school reducers here, won't this obliterate any persistence for the modularized reducers?

By "obliterating any persistence", are you referring to serialization or deserialization?

As I understand, what will happen during deserialization is that the legacy reducers will all get initialized in one go, when booting. Modularized reducers will get initialized later, as they're loaded. So there should be no issues on deserialization, as we'll still load all the data that's available to be loaded.

Serialization is different: the only things that get serialized are the ones that are loaded. So, once state is fully modularized, if a user is on a route with few reducers and state gets persisted, very little information will be written to storage. If that's indeed a concern, we could fix it by e.g. storing each top-level reducer in its own storage key.

jsnajdr commented:
> Looks like matticbot is producing incorrect numbers;

That happens when the `master` commit your branch is based on doesn't have ICFY stats. In the last week, there are 4 `master` commits that failed to build `icfy-stats` 😢 I'll need to look into this.

sgomes commented:
> That happens when the `master` commit your branch is based on doesn't have ICFY stats. In the last week, there are 4 `master` commits that failed to build `icfy-stats` 😢 I'll need to look into this.

Thank you, @jsnajdr! In the meantime, would it be easier to perhaps detect when the error occurs, and stop the bad numbers from being displayed?

sgomes commented:
Rebased on master to attempt to pick up a good branch state for ICFY.

sgomes commented:
Now that we have some correct bundle numbers, the results are quite interesting. It appears that we save around 5KB (compressed) in the main entry point, but we introduce some duplication in the section chunks.

Much of the change appears to be things shifting around based on heuristics, but it appears the overall trend is positive, at least as far as I understand.

dmsnell commented:
It's surprising to see the `login` entry point grow - is that a quirk or is it more simply a byproduct of some hard `import` on the reader vs. an async import?

dmsnell commented:
> Breaking modularized reducers out to use their own key would indeed solve the issue. As long as we're modularizing at the top level, that should be somewhat straight-forward.

Seems like this is a win anyway to split one giant atomic update in `indexedDB` to many smaller ones.

jsnajdr commented on client/state/reader/reducer.js:
```diff
@@ -36,3 +37,6 @@ export default combineReducers( {
 } );
+export default combinedReducer;
+
+registerReducer( [ 'reader' ], combinedReducer );
```
One idea how to do the reducer declaration a bit better:
- keep the `reducer.js` module side-effect-free, just exporting just the reducer. That's nice for unit testing it
- create a new file, `state/reducer/index.js`, which does the registration:
```js
import reader from './reducer';
registerReducer( 'reader', reader );
```
Then, all action files that want to trigger the registration would:
```js
import 'state/reader';
```
without the `/reducer` suffix.

As an API, I like this much better.

jsnajdr replied:
You know what would be even more awesome? If, with some `package.json` magic, we could ensure that any import from `state/reader/*` automatically ensures that the reducer is registered. Think about DLL libraries and their `LoadLibrary` and `FreeLibrary` functions.

sgomes replied:
> You know what would be even more awesome? If, with some `package.json` magic, we could ensure that any import from `state/reader/*` automatically ensures that the reducer is registered. Think about DLL libraries and their `LoadLibrary` and `FreeLibrary` functions.

Interesting! I'll look into whether this is easily doable. Then we'd only have to manually do this for selectors and action creators that touch multiple top-level branches of state.

jsnajdr replied:
> I'll look into whether this is easily doable.

I'm afraid it won't be feasible at this moment. We'd have to create a top-level module with initialization and reexports:
```js
import './init';
export { default as reducer } from './reducer';
export { ... } from './actions';
export { ... } from './selectors';
```
The most feasible way to make progress here seems to be packaging it all in a `@wordpress/data` store.

sgomes replied:
I've been looking at the webpack side of things and there doesn't seem to be anything there either. We'd need to write and maintain our own webpack plugin for this. And I do think that being explicit in the dependency graph is probably better.

tyxla replied:
Just saying that this new API is beautiful. Over time we can improve/automate the reducer registration, but I think that this is already a huge leap forward!

jsnajdr commented:
To make state persistence work, you need to declare the Reader reducer with `withStorageKey`:
```js
const readerReducer = combineReducer( { conversations, feeds, ... } );
export default withStorageKey( 'reader', readerReducer );
```

If you don't do this, persisted state will be lost when navigating (with reload) between different sections that either have `state.reader` loaded or not.

When loading persisted state with `state.reader` into a reducer without `reader`, the state will not be loaded into memory -- you need a working reducer for that. When later writing the state to storage, it will be written without the `reader` subtree, losing the previously stored `reader` state.

When using `withStorageKey`, the `reader` state will be stored into a separate IDB key. And then loaded again either [at boot](https://github.com/Automattic/wp-calypso/blob/master/client/state/initial-state.js#L264-L283) or when [dynamically loading](https://github.com/Automattic/wp-calypso/blob/master/client/state/add-reducer.ts#L60) the reducer.

I think we should enforce that when a reducer:
- supports persistence (`SERIALIZE` returns something other than `undefined`)
- is dynamically loaded
then it **must** define `reducer.storageKey`. If it doesn't, persistence is guaranteed to not work.

jsnajdr commented:
Seeing how many selector files had to be changed, it turns out that organizing the selectors in one big flat `state/selectors` directory, one selector per file, was not a good idea. At least for this case.

By modularizing the reducers, we are abandoning the principle that Redux actions, reducers and selectors are completely decoupled. That an action creator doesn't need to know which reducers react to it (or in other words, that an event publisher doesn't need to be aware of the event subscribers). And that a selector can freely select from any part of the state it wishes.

By now, the reducer, the actions and the selectors are coupled tightly into one `state/reader` module. Therefore, all the Reader selectors should be moved to `state/reader/selectors`.

Then we could apply a rule that an action or selector needs to `import 'state/reader/reducer` if and only if it lives inside `state/reader/{actions,selectors}`.

sgomes commented:
Thank you, @blowery and @jsnajdr for your great suggestions!

As you point out, @jsnajdr, we are adding existing restrictions on top of the Redux architecture. We're requiring developers to explicitly link a selector or action creator to the initialisation of the reducers that manage the parts of state it depends on.

I'm guessing we'll definitely have action creators or selectors that need to interact with multiple parts of state, but that's a simple matter of keeping only those in something like `state/selectors`. And it's functionally easy to make them work: they just have to import the inits for all of the reducers they depend on. 

```js
// omni-selector.js
import 'state/reader/init';
import 'state/sites/init';
```

As you may guess from the above, taking your idea, the only change I'd make is naming the side-effect module `init` instead of `index` to make it clearer. Since most files accessing it are going to be living within the directory structure anyway, it doesn't feel like a huge violation of abstraction.

So, I can think of two alternatives on how to organise things. Here's alternative A:

```
state/reader
  |- init.js // this loads `reducer/index` and registers it
  |
  |- reducer
  |  |- index.js
  |  |- feeds/
  |  |- sites/
  |  |- [...]
  |
  |- selectors
  |  |- feeds/index.js // depends on `state/reader/init`
  |  |- sites/index.js // depends on `state/reader/init`
  |  |- [...]
  |
  |- actions
  |  |- feeds/index.js // depends on `state/reader/init`
  |  |- sites/index.js // depends on `state/reader/init`
  |  |- [...]
```

The above "flattens" all of reader state. It creates an identical structure under `reducer`, `selectors` and `actions`, with each of the sub-structures of reader in there (sites, feeds, etc.).

Another is creating a `reducer`, `selectors` and `actions` inside each sub-section of the tree. Let's call it alternative B:

```
state/reader
  |- init.js     // this loads `reducer.js` and registers it
  |- reducer.js  // combines feeds/reducer, sites/reducer, etc.
  |
  |- feeds/
  |  |- selectors/index.js // depends on `state/reader/init`
  |  |- actions/index.js   // depends on `state/reader/init`
  |  |- reducer.js
  |
  |- sites/
  |  |- selectors/index.js // depends on `state/reader/init`
  |  |- actions/index.js   // depends on `state/reader/init`
  |  |- reducer.js
```

This emphasizes the tree-like nature of state, but it splits selectors, actions and reducers across the board. And there would likely still be a need for some selectors and actions at the top level, if they affect multiple sub-sections.

I'm leaning towards alternative A, but which do you think is best?

jsnajdr commented:
> I'm leaning towards alternative A, but which do you think is best?

I don't have any strong opinion here. The `state/reader` folder tree is actually a two-dimensional structure, where `{reducer,selectors,actions}` is on one axis and `{feeds,sites,....}` is on a second axis. Which one is first and second when flattening is completely arbitrary.

The current `state/reader` tree follows the alternative B. So, if I were you, I would leave it as is without a massive reorg.

jsnajdr commented:
> As you may guess from the above, taking your idea, the only change I'd make is naming the side-effect module init instead of index to make it clearer.

That sounds great. `init` is better than `index` 🙂

sgomes commented:
> I don't have any strong opinion here. The `state/reader` folder tree is actually a two-dimensional structure, where `{reducer,selectors,actions}` is on one axis and `{feeds,sites,....}` is on a second axis. Which one is first and second when flattening is completely arbitrary.
> 
> The current `state/reader` tree follows the alternative B. So, if I were you, I would leave it as is without a massive reorg.

That's a good point. Reorganising everything would be painful. Let's go with B, then!

sgomes committed:
Move special deserialize handling to reducer-utils

sgomes commented on client/state/initial-state.js:
```diff
@@ -40,6 +39,14 @@ function serialize( state, reducer ) {
+	// more properties than the reducers that have been loaded so far.
+	if ( reducer.keys ) {
+		return reducer( pick( state, reducer.keys ), { type: DESERIALIZE } );
+	}
```
That's an excellent idea, thank you! This is so much cleaner! 😃

sgomes commented:
> To make state persistence work, you need to declare the Reader reducer with `withStorageKey`

That's perfect, thank you! I hadn't thought of using storage keys.

I think adding a storage key to each modularised reducer is a much better alternative to making further changes to the persistence system 👍

sgomes committed:
Add storage key to modularized reducer.

sgomes committed:
Move registration side-effect to its own module

sgomes committed:
Move getBlockedSites to state/reader

sgomes committed:
Move getSiteBlocksCurrentPage to state/reader

sgomes committed:
Move getSiteBlocksLastPage to state/reader

sgomes committed:
Move isFetchingSiteBlocks to state/reader

sgomes committed:
Move isSiteBlocked to state/reader

sgomes commented:
@blowery @jsnajdr : Should I continue moving reducers into `state/reader/` as part of this PR, or should I do it separately, after this PR gets merged?

sgomes commented:
> It's surprising to see the `login` entry point grow - is that a quirk or is it more simply a byproduct of some hard `import` on the reader vs. an async import?

@dmsnell I looked into why this is the case, with some bundle analysis. Looks like it's due to `initial-state` now being needed, transitively, because of the dependency on `state/redux-store` and `state/add-reducer`.

We could drop the changes from `entry-login`, since we manually craft a smaller set of reducers there anyway, but ultimately I'd like to replace that manual set of reducers with the same automatic modularised reducer mechanism we're using in `entry-main`.

sgomes committed:
Explicitly list side effects in state/reader

sgomes commented:
I think I've addressed everyone's comments and made all requested changes at this point, save for completing the migration of reader-related reducers into `state/reader`. Please let me know if I missed something!

jsnajdr commented on client/reader/following-manage/index.jsx:
```diff
@@ -17,7 +17,7 @@ import DocumentHead from 'components/data/document-head';
 import HeaderCake from 'components/header-cake';
 import ReaderMain from 'reader/components/reader-main';
-import getBlockedSites from 'state/selectors/get-blocked-sites';
+import getBlockedSites from 'state/reader/site-blocks/selectors/get-blocked-sites';
```
I'd prefer to do my imports as:
```js
import { getBlockedSites } from 'state/reader/site-blocks/selectors';
```
i.e., from a single umbrella `*/selectors` module.

sgomes replied:
Good idea! 👍 I added an index file that re-exports all the selectors, and updated all the imports to use that instead.

jsnajdr reviewed (approved):
This looks great! :tada:

sgomes committed:
Add index file to site-blocks selectors.

tyxla commented on client/state/reader/package.json:
```diff
@@ -0,0 +1,5 @@
+{
+	"sideEffects": [
```
Nice 👍

tyxla reviewed (approved):
Also, I ran it through some tests, and couldn't break it. Stellar work 🚢

sgomes commented:
Thanks for the reviews, everyone! 👍
